### ISSUE TITLE
Report inference origin as a sub-diagnostic in (some?) failed assignability cases

### ISSUE THREAD

**Issue by sahilmtayade** [astral-sh/ty#4056]
**Reproduced on:** play.ty.dev  
**Python version:** 3.x

---

### Summary

When a variable is initialised as `list[int]` and later reassigned to `list[float]` inside a loop,
`ty` reports the error at the `return` statement — several lines away from where the type was
actually fixed. This is a red herring. The real problem is at the initialisation site, and `ty`
never points there.

---

### Reproduction

**Playground:** https://play.ty.dev/c83f5c8f-197e-46cc-8187-e18b05ddb4f8

```python
from typing import Sequence

def _sum_monthly_vectors(vector_a: Sequence[float], vector_b: Sequence[float]) -> list[float]:
    return [a + b for a, b in zip(vector_a, vector_b, strict=True)]


def _collect(parent_data) -> list[float]:
    v3 = 12 * [0]          # ← root cause: inferred as list[int]

    for sub_data in parent_data:
        v1 = [3.3] * 12
        v3 = _sum_monthly_vectors(v3, v1)

    return v3              # ← error reported here, far from the actual problem
```

**Error reported:**

```
Return type does not match returned value: expected `list[int | float]`, found `list[int] | list[int | float]`  (invalid-return-type) [Ln 13, Col 12]

    Expected `list[int | float]` because of return type
    info: element `list[int]` of union `list[int] | list[int | float]` is not assignable to `list[int | float]`
    info: └── element `float` of union `int | float` is not assignable to ...
```

---

### Why this is a bad diagnostic

The error message surfaces a `list[int] | list[int | float]` union — an artefact of control-flow
analysis across the loop — without ever mentioning that `v3 = 12 * [0]` is what introduced
`list[int]` in the first place.

A developer reading this error will naturally question the return type annotation or the function
signature. Almost no one will look back at what appears to be a routine accumulator initialisation.
The actual fix is simple and obvious — once you know where to look:

```python
v3 = [0.0] * 12           # list[float] ✅
# or
v3: list[float] = [0] * 12  # explicit annotation ✅
```

But `ty` never tells you that.

For comparison, the **identical logic without a loop** passes without any error:
https://play.ty.dev/3aae852d-5b64-4b21-b078-80a6de512b6a

This means the confusing union in the error message is a loop control-flow artefact, making the
diagnostic even harder to interpret.

---

### What a good diagnostic would look like

When `ty` detects a conflict traceable to an earlier inference site, it should point there — similar
to how `rustc` traces type mismatches back to where a type was first inferred:

```
error: return type does not match returned value  [invalid-return-type] [Ln 13]
    |
  8 |     v3 = 12 * [0]
    |          ^^^^^^^^ `v3` was inferred as `list[int]` here
    |
    = note: if a `list[float]` was intended, consider `[0.0] * 12`
            or annotate explicitly: `v3: list[float] = [0] * 12`
```

Pointing to the initialisation site, and suggesting the fix there, would turn an opaque union
error into an immediately actionable message.

---

### Impact

The pattern of initialising a numeric accumulator with `[0] * n` before a loop is extremely
common in data and scientific code. The current diagnostic consistently misdirects developers
toward the wrong part of the code. Improving it would meaningfully reduce confusion for this
whole class of errors.

---
Disclaimer: AI was used to format this issue request.

**Comment by carljm** [comment]
Thanks. I agree that we would ideally point to the origin of the type inference, particularly in a case like this where two different inference origins result in a union type. I think it may be tricky to identify the precise heuristic that should drive when additional detail is provided.

**Comment by carljm** [comment]
The difference between the loop and straight-line version is easy to explain: in the straight line version there is no possibility for the initial list to reach the return. In the loop version this is possible, because the loop may execute zero times (we have no basis for knowledge of whether `parent_data` is empty or not.)
